tsdb/exemplar.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tsdb
+
+import (
+	"context"
+	"sync"
+
+	"github.com/prometheus/prometheus/pkg/exemplar"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+)
+
+// single circular buffer for all exemplars
+type CircularExemplarStorage struct {
+	lock        sync.RWMutex
+	index       map[string]int
+	exemplars   []*circularBufferEntry
+	nextIndex   int
+	len         int
+	secondaries []storage.ExemplarAppender
+}
+
+type circularBufferEntry struct {
+	se   storageExemplar
+	prev int // index of previous exemplar in circular for the same series, use -1 as a default for new entries
+}
+
+func (ce circularBufferEntry) Exemplar() exemplar.Exemplar {
+	return ce.se.exemplar
+}
+
+type storageExemplar struct {
+	exemplar        exemplar.Exemplar
+	seriesLabels    labels.Labels // need to store labels so we can double check when querying
+	scrapeTimestamp int64
+}
+
+// If we assume the average case 95 bytes per exemplar we can fit 5651272 exemplars in
+// 1GB of extra memory, accounting for the fact that this is heap allocated space.
+func NewCircularExemplarStorage(len int, secondaries ...storage.ExemplarAppender) *CircularExemplarStorage {
```

Comment:
should the exemplar storage take secondaries, or should there be a new `ExemplarFanout` in storage?